1.20+ Advancements Support

VDPineapple opened the pull request:
![ninbotexample](https://github.com/user-attachments/assets/c3854c1e-2528-4090-960d-e80db30afcf3)

Added the following features:
- New setting to Optional features > All Advancements window: 1.20+ All Advancements Mode with more rows for 1.20+ features.
    - Stronghold: Unchanged
    - Spawn: Unchanged
    - Outpost: Activates with F3+C when 63 >= Y < 160
    - Monument: Activates with F3+C when 30 > Y < 63
    - Shrieker: Activates with F3+C when Y <= 30. 1.20+ runs need to transport a warden back to spawn.
    - End City: Activates with F3+C when in the end. 1.21 runs usually transport shulker after initial end entry with leading boats.
    - Ancient City Region Centres: Activates with F3+C when Y > 160. Ancient city regions are centred around 24m+8 chunk coordinates, where m is an integer. This helps find the closest ancient city region centre to improve dig-down odds.
    - General Location: Activates with F3+I when all advancements mode is toggled. Useful when you need a temporary coordinate written down which you will return to in the short-term. Typical scenarios include ancient city dig-down spots, tadpole spawning spot, foxes.
    
Known side effects:
- New MCDimension class instead of inNether booleans.
- To make space for the new rows, allowed the allAdvancements panel to use its preferred size. This changes the regular allAdvancements panel height, making it slightly bigger.
- Not much space to add the instructions for when each structure-type triggers, so there are no new instructions in the program itself.
- Due to the length of 1.20+ runs, they are often done in multiple segments, which will wipe the bot. It may be helpful to write coordinates in a file in the world folder at some point in the future.

Ninjabrain1 commented on src/main/java/ninjabrainbot/model/datastate/endereye/CoordinateInputSource.java:
```diff
@@ -53,6 +60,13 @@ private void parseF3C(String f3c) {
 			whenNewFossilInputted.setAndAlwaysNotifySubscribers(f);
 		}
+
+		if (preferences.oneDotTwentyPlusAA.get() && preferences.allAdvancements.get()) {
```
Few comments about this:

- AA preferences should not matter here. This class's only purpose is to provide the coordinates the user inputs, it should not care about the AA preferences. setAndAlwaysNotifySubscribers should always be called whenever a new F3+I is detected, but the subscriber should then determine whether or not to do something (or ignore the F3+I) based on the AA preferences.
- Parsing should not be done in the StructurePosition class, CoordinateInputSource is a general class for providing what the user inputs, its too specific to have StructurePosition code here. Instead create a class similar to F3CData, and do F3IData.tryParseF3IString(f3c);
- I do not like the name "GeneralLocation" because it refers to the specific usage in the new AA mode. I think a better name is just F3ILocation. Rename this everywhere, e.g. GeneralLocationInputHandler -> F3ILocationInputHandler.

Ninjabrain1 replied:
Btw you can keep the name "GeneralLocation" in the places where it specifically referrs to the general location used in AA mode, such as in IAllAdvancementsDataState

Ninjabrain1 commented on src/main/java/ninjabrainbot/model/datastate/endereye/F3CData.java:
```diff
@@ -23,16 +23,21 @@ public static F3CData tryParseF3CString(String string) {
 			String world = substrings[2];
-			if (world.endsWith("_end")) {
-				return null;
+			MCDimension dimension = null;
```
Refactor this into its own method getMCDimension

Ninjabrain1 commented on src/main/java/ninjabrainbot/model/actions/common/SetGeneralLocationAction.java:
```diff
@@ -0,0 +1,22 @@
+    private final StructurePosition structurePosition;
+
+	public SetGeneralLocationAction(IAllAdvancementsDataState allAdvancementsDataState, StructurePosition structurePosition) {
+        this.allAdvancementsDataState = allAdvancementsDataState;
```
Indentation

Ninjabrain1 commented on src/main/java/ninjabrainbot/model/actions/common/SetGeneralLocationAction.java:
```diff
@@ -0,0 +1,22 @@
+
+	@Override
+	public void execute() {
+        allAdvancementsDataState.generalLocationPosition().set(structurePosition);
```
indentation

Ninjabrain1 commented on src/main/java/ninjabrainbot/model/actions/common/SetGeneralLocationAction.java:
```diff
@@ -0,0 +1,22 @@
+package ninjabrainbot.model.actions.common;
```
move to alladvancements package (instead of common)

Ninjabrain1 commented on src/main/java/ninjabrainbot/model/actions/alladvancements/TryAddAllAdvancementsStructureAction.java:
```diff
@@ -28,29 +47,49 @@ public void execute() {
 	private StructureType getAllAdvancementStructureTypeFromPlayerPosition(IDetailedPlayerPosition t) {
-		if (t.isInNether())
-			return StructureType.Unknown;
+		if (!preferences.oneDotTwentyPlusAA.get()) {
```
There is too much nesting and too much repetition in this method. I think this would achieve the same result but with less repetition:

1. If in nether, return Unknown (this applies to both pre and post 1.20)
2. If in the end, return ShulkerTransport if 1.20+, or Unknown if pre 1.20
3. Check for spawn
4. Check for monument
5. If 1.20+, check for DeepDark and CityQuery
6. Return Outpost

Ninjabrain1 commented on src/main/java/ninjabrainbot/model/actions/alladvancements/TryAddAllAdvancementsStructureAction.java:
```diff
@@ -12,14 +13,32 @@
+		}
+	}
+
+	public void generalLocationSet(StructurePosition structurePosition) {
```
Remove, generalLocationSet is unused right?

Ninjabrain1 commented on src/main/java/ninjabrainbot/model/actions/alladvancements/TryAddAllAdvancementsStructureAction.java:
```diff
@@ -12,14 +13,32 @@
+		this.preferences = preferences;
+	}
+
+	public StructurePosition getStructurePosition(StructureType structureType) {
```
Make this private

VDPineapple commented:
Thanks for your feedback, I've fixed all of the things you mentioned. I also fixed the remove button for the new structures. Let me know if you'd like any further changes.
